rust/lance-index/src/scalar/ngram.rs

```diff
@@ -0,0 +1,456 @@
+                let row_ids = NGramPostingList::intersect(list_refs);
+                Ok(SearchResult::AtMost(RowIdTreeMap::from(row_ids)))
+            }
+        }
+    }
+
+    fn can_answer_exact(&self, _: &dyn AnyQuery) -> bool {
+        false
+    }
+
+    /// Load the scalar index from storage
+    async fn load(store: Arc<dyn IndexStore>) -> Result<Arc<Self>>
+    where
+        Self: Sized,
+    {
+        Ok(Arc::new(Self::from_store(store.as_ref()).await?))
+    }
+
+    /// Remap the row ids, creating a new remapped version of this index in `dest_store`
+    async fn remap(
+        &self,
+        _mapping: &HashMap<u64, Option<u64>>,
+        _dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        todo!()
+    }
+
+    /// Add the new data into the index, creating an updated version of the index in `dest_store`
+    async fn update(
+        &self,
+        _new_data: SendableRecordBatchStream,
+        _dest_store: &dyn IndexStore,
+    ) -> Result<()> {
+        todo!()
+    }
+}
+
+pub struct NGramIndexBuilder {
+    tokenizer: TextAnalyzer,
+    tokens_map: HashMap<String, u32>,
+    tokens_counter: u32,
+    bitmaps: Vec<RoaringTreemap>,
+}
+
+impl Default for NGramIndexBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl NGramIndexBuilder {
+    pub fn new() -> Self {
+        let tokenizer = NGRAM_TOKENIZER.clone();
+        Self {
+            tokenizer,
+            // Default capacity loosely based on case insensitive ascii ngrams with punctuation stripped
+            tokens_map: HashMap::with_capacity(4 * 26 * 26 * 26),
+            bitmaps: Vec::with_capacity(4 * 26 * 26 * 26),
+            tokens_counter: 0,
+        }
+    }
+
+    fn validate_schema(schema: &Schema) -> Result<()> {
+        if schema.fields().len() != 2 {
+            return Err(Error::InvalidInput {
+                source: "Ngram index schema must have exactly two fields".into(),
+                location: location!(),
+            });
+        }
+        if *schema.field(0).data_type() != DataType::Utf8 {
+            return Err(Error::InvalidInput {
+                source: "First field in ngram index schema must be of type Utf8".into(),
+                location: location!(),
+            });
+        }
+        if *schema.field(1).data_type() != DataType::UInt64 {
+            return Err(Error::InvalidInput {
+                source: "Second field in ngram index schema must be of type UInt64".into(),
+                location: location!(),
+            });
+        }
+        Ok(())
+    }
+
+    fn process_batch(&mut self, batch: &RecordBatch) {
+        let text_col = batch.column(0).as_string::<i32>();
+        let row_id_col = batch.column(1).as_primitive::<UInt64Type>();
+        for (text, row_id) in text_col.iter().zip(row_id_col.values()) {
+            let text = text.unwrap();
+            let tokens = self.tokenizer.clone().tokenize(text);
```

Comment:
should this list be deduplicated? I don't see any indication tantivy handles that for us

Reply:
I also wonder if during the build step, there would be any value in caching computed trigrams and potentially splitting the input to try and get better utility out of the cache (e.g for words).

Reply:
one more thing to pile on - I was curious how expensive the per-record allocations for token vectors are in the build phase. Does that consume meaningful time vs reusing a vector?

Reply:
>  I was curious how expensive the per-record allocations for token vectors are in the build phase. Does that consume meaningful time vs reusing a vector?

It could probably be substantial.  I've removed the excess allocations.

> should this list be deduplicated? I don't see any indication tantivy handles that for us

Tantivy does not deduplicate.  I'm not sure how much savings we would get in training.  We'd have to go back to allocating the vector so that we could sort the tokens.  We can test it later.

It might be better to deduplicate on search, as I think right now we end up doing a bitmap AND of a treemap with itself which isn't trivial.

Reply:
@westonpace I did a bit of experimentation with the patch on a dataset of reddit comments. My findings were,
* No savings from sort + dedup - works out slower. I think probably inserting "duplicates" into the bitmap is very quick.
* ~25% speedup in the build from a caching strategy, where I split the input strings on spaces, tokenize each "word" individually, and maintain a cache of word tokenizations (hash map). I didn't verify that this results in the same tokenizations as tokenizing the full input in one shot but I think it probably would? Might be something to explore.

My biggest concern with the patch is super long build times but I think the biggest win on that will probably be parallelization at this point.

Reply:
> My biggest concern with the patch is super long build times but I think the biggest win on that will probably be parallelization at this point.

Yep.  Just testing now on 200M large docs and it's pretty brutal.  I'll try and get a follow-up with parallelization quickly.